l3kernel/l3bitset.dtx

```diff
@@ -0,0 +1,551 @@
+                \str_put_left:Nn #1 { 0 }
+              }
+            \str_put_left:Nn #1 { 1 }
+          }
+          {
+            %replace value
+            \str_set:Nx #1
+              {
+                \str_range:Nnn #1 {1}{-2 - (#2)}
+                1
+                \str_range:Nnn #1 {0 -(#2)}{-1}
+             }
+          }
+      }
+  }
+
+%#1 name, #2 index (integer expression,zero based)
+\cs_new_protected:Npn \bitset_gset:Nn #1 #2
+  {
+    \int_compare:nNnT {#2 } > { -1 }
+      {
+        \int_compare:nNnTF {\str_count:N #1 } < { #2 + 1 }
+          {
+            %extend the str
+            \int_step_inline:nn { #2 - \str_count:N #1 }
+              {
+                \str_gput_left:Nn #1 { 0 }
+              }
+            \str_gput_left:Nn #1 { 1 }
+          }
+          {
+            % replace the value
+            \str_gset:Nx #1
+              {
+                \str_range:Nnn #1 {1}{-2 -(#2)}
+                1
+                \str_range:Nnn #1 {0 -(#2)}{-1}
+              }
+          }
+      }
+  }
+\cs_generate_variant:Nn \bitset_set:Nn  {cn}
+\cs_generate_variant:Nn \bitset_gset:Nn {cn}
+%    \end{macrocode}
+% \end{macro}
+
+% \begin{macro}
+%   {
+%     \bitset_clear:Nn,  \bitset_clear:cn,
+%     \bitset_gclear:Nn, \bitset_gclear:cn
+%   }
+% Clearing a bit has only to do something if the string is longer than then index.
+%    \begin{macrocode}
+\cs_new_protected:Npn \bitset_clear:Nn #1 #2  %#1 name, #2 index (zero based)
+ {
+   \int_compare:nNnT {#2 } > { -1 }
+     {
+       \int_compare:nNnT {\str_count:N #1 } > { #2 }
+         {
+           % need to replace the str
+           \str_set:Nx #1
+             {
+               \str_range:Nnn #1 {1}{-2 - (#2)}
+                0
+               \str_range:Nnn #1 { 0 - (#2)}{-1}
+             }
+         }
+     }
+ }
+
+\cs_new_protected:Npn \bitset_gclear:Nn #1 #2  %#1 name, #2 index (zero based)
+ {
+   \int_compare:nNnT {#2 } > { -1 }
+     {
+       \int_compare:nNnT {\str_count:N #1 } > { #2 }
+         {
+           % need to replace the str
+           \str_gset:Nx #1
+             {
+               \str_range:Nnn #1 {1}{-2 -(#2)}
+                0
+               \str_range:Nnn #1 {0 - (#2)}{-1}
+             }
+         }
+     }
+ }
+\cs_generate_variant:Nn \bitset_clear:Nn {cn}
+\cs_generate_variant:Nn \bitset_gclear:Nn {cn}
+%    \end{macrocode}
+%  \end{macro}
+% \begin{macro}
+%   {
+%     \bitset_to_arabic:N, \bitset_to_arabic:c,
+%     \bitset_to_bin:N,    \bitset_to_bin:c,
+%   }
+%   The naming of the commands follow the names in the \texttt{int} module.
+%   \cs{bitset_to_arabic:N} uses \cs{int_from_bin:n} and so can only handle small
+%   numbers. There is no test for the string length, so that long bitset with
+%   many leading zeros work nevertheless.
+%    \begin{macrocode}
+\cs_new:Npn \bitset_to_arabic:N #1
+  {
+    \exp_args:Ne \int_from_bin:n { #1 }
+  }
+
+\cs_new:Npn \bitset_to_bin:N #1
+  {
+    #1
+  }
```

Comment:
Especially when showing a bitset to the user, in my experience you either want a trimmed bitset or a bitset of a predefined constant width. So maybe we could add a `\bitset_to_bin:Nn` which extends the bitset to at least `#2` digits?

Reply:
I think the 'defined width' part is a more general point. Should the set have a pre-defined fixed length? This would address the question about what a the 'cleared' set looks like. Might also be useful as at present one can't predict how long the set will be if you dump it as a binary, which would be 'fun' if it was going somewhere you then wanted to pull apart on a bit basis.

Reply:
well I don't think about bitset in this way. For me it is a set of "set of indexed bits". The length would be for me the highest bit set. Zeros are for me "implementation detail", and  000, 00, 0 and "empty" are the same bitset, only with a different representation.